File: 2020/cpp/day_07.cpp
#include "day_07.h"

day_07::day_07() : input_strings{readFile("Inputs/day_07.txt")} {
    const auto start = std::chrono::high_resolution_clock::now();
    for (const auto& line : input_strings) {
        auto words = split(line, ' ');
        const std::string_view colour = {&line.front(), &(words[1].back()) + 1};
        rules[colour] = {};
        for (auto i = words.begin() + 4; i < words.end(); i += 4) {
            auto& pair = rules[colour].emplace_back();
            pair.colour = {&i[1].front(), &(i[2].back()) + 1};
            pair.number = fast_atol(i[0].data(), 1);
            fmt::print("");
        }
        bags.emplace_back(colour);
    }
    std::ranges::sort(bags);
    bags.erase(std::unique(bags.begin(), bags.end()), bags.end());

    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("Parsing input for day six took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_07::part_one() {
    const auto start = std::chrono::high_resolution_clock::now();
    int count{0};
    for (auto colour : bags) {
        if (colour == "shiny gold")
            continue;
        std::vector<std::string_view> candidates;
        candidates.reserve(rules.size());
        candidates.push_back(colour);
        while (!candidates.empty()) {
            auto bag = candidates.front();
            if (bag == "shiny gold") {
                ++count;
                break;
            }
            for (const auto& subbag : rules[bag]) {
                candidates.emplace_back(subbag.colour);
            }
            pop_front(candidates);
        }
    }
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day seven part one is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_07::part_two() {
    const auto start = std::chrono::high_resolution_clock::now();
    int count{0};
    std::vector<rule_t> candidates;
    candidates.push_back({"shiny gold", 1});
    while (!candidates.empty()) {
        const auto& bag = candidates.front();
        count += bag.number;

        for (const auto& subbag : rules[bag.colour]) {
            candidates.emplace_back(subbag.colour, subbag.number * bag.number);
        }
        pop_front(candidates);
    }
    --count;  // we need to remove the original shiny gold bag
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day seven part two is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

File: 2020/Rust/src/day07.rs
use std::collections::HashMap;
use std::collections::VecDeque;
use std::time::Duration;
use std::time::Instant;

#[derive(Debug, Clone)]
struct Rule<'a> {
    colour: &'a str,
    number: usize,
}

fn parse_input(input: &str) -> ((HashMap<&str, Vec<Rule<'_>>>, Vec<&str>), Duration) {
    let start = Instant::now();
    let mut bags: Vec<&str> = Vec::new();
    let mut rules: HashMap<&str, Vec<Rule>> = HashMap::new();
    for line in input.lines() {
        let split = line.split(' ').collect::<Vec<_>>();
        let colour = unsafe {
            std::str::from_utf8_unchecked(std::slice::from_raw_parts(
                line.as_ptr(),
                split[2].as_ptr().sub(1) as usize - line.as_ptr() as usize,
            ))
        };
        let mut i = 4;
        loop {
            if i >= split.len() {
                break;
            }
            let v = rules.entry(colour).or_default();
            if let Ok(number) = split[i].parse() {
                v.push(Rule {
                    colour: unsafe {
                        std::str::from_utf8_unchecked(std::slice::from_raw_parts(
                            split[i + 1].as_ptr(),
                            split[i + 2].as_ptr().add(split[i + 2].len() - 2) as usize
                                - split[i].as_ptr() as usize,
                        ))
                    },
                    number,
                })
            }
            i += 4;
        }
        bags.push(colour)
    }
    let end = Instant::now();
    ((rules, bags), end - start)
}

fn solve_part1(rules: &HashMap<&str, Vec<Rule>>, bags: &[&str]) -> (usize, Duration) {
    let start = Instant::now();
    let mut count = 0;
    for &colour in bags {
        if colour == "shiny gold" {
            continue;
        }
        let mut candidates = VecDeque::with_capacity(rules.len());
        candidates.push_back(colour);
        while let Some(&bag) = candidates.front() {
            if bag == "shiny gold" {
                count += 1;
                break;
            }
            for subbag in &rules[bag] {
                candidates.push_back(subbag.colour);
            }
            candidates.pop_front();
        }
    }
    let end = Instant::now();
    (count, end - start)
}

fn solve_part2(rules: &HashMap<&str, Vec<Rule>>) -> (usize, Duration) {
    let start = Instant::now();
    let mut count = 0;
    let mut candidates: VecDeque<Rule> = VecDeque::new();
    candidates.push_back(Rule {
        colour: "shiny gold",
        number: 1,
    });
    while let Some(rule) = candidates.front().cloned() {
        count += rule.number;

        for subbag in &rules[rule.colour] {
            candidates.push_back(Rule {
                colour: subbag.colour,
                number: subbag.number * rule.number,
            });
        }
        candidates.pop_front();
    }
    count -= 1;
    let end = Instant::now();
    (count, end - start)
}

pub fn run(input: String) {
    let ((rules, bags), time) = parse_input(&input);
    println!("Day 07, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&rules, &bags);
    println!("Day 07, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&rules);
    println!("Day 07, part 2: {}. Took {}ns", p2, time.as_nanos());
}
